### PR TITLE
fix(iceberg): fix map - arrow convert & fix sink nested types to iceberg

### LINKED ISSUE #16545: iceberg sink: struct type with metadata doesn't work
Hi, I've recently been thinking about support for Struct type in Iceberg sink, since I'm testing if I can utilise RisingWave at work and such functionality is a necessity.  As of now when someone tries to sink struct data to iceberg catalog they receive an error
```Field response's type not compatible, risingwave converted data type Struct([Field { name: "responseStatus", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "statusCode", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }]), iceberg's data type: Struct([Field { name: "responseStatus", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {"PARQUET:field_id": "17", "column_id": "17"} }, Field { name: "statusCode", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {"PARQUET:field_id": "18", "column_id": "18"} }])```
Looking at this it seems it is only a matter of a mismatch on `metadata` field in each `Field`. The code just does `left == right` comparison <https://github.com/risingwavelabs/risingwave/blob/main/src/connector/src/sink/iceberg/mod.rs#L1046>

Is Struct support in Iceberg sink just a matter of lack of correct comparison or is there more context to that?

[Slack Message](https://risingwave-community.slack.com/archives/C02T3F7UYM6/p1714404363778909?thread_ts=1714404363.778909&cid=C02T3F7UYM6)

### LINKED ISSUE #18451: bug: map type has been converted into an arrow map type with a wrong field type
### Describe the bug

The inner field data type of arrow map type should be a `Struct<k,v>` instead of `List<Struct<k,v>>`
https://github.com/risingwavelabs/risingwave/blob/c843edcd298be55a0bdc938dfa3c240676fba9c1/src/common/src/array/arrow/arrow_impl.rs#L449-L460

> /// A Map is a logical nested type that is represented as
    ///
    /// `List<entries: Struct<key: K, value: V>>`
    ///
    /// The keys and values are each respectively contiguous.
    /// The key and value types are not constrained, but keys should be
    /// hashable and unique.
    /// Whether the keys are sorted can be set in the `bool` after the `Field`.
    ///
    /// In a field with Map type, the field has a child Struct field, which then
    /// has two children: key type and the second the value type. The names of the
    /// child fields may be respectively "entries", "key", and "value", but this is
    /// not enforced.

### Error message/log

_No response_

### To Reproduce

_No response_

### Expected behavior

_No response_

### How did you deploy RisingWave?

_No response_

### The version of RisingWave

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
- fix #18451
- fix #16545
- Note: source with nested type is not supported yet, because not implemented in iceberg-rs #18474
- Note: create table in risingwave side is not supported yet https://github.com/risingwavelabs/risingwave/issues/18363#issuecomment-2340432750

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/connector/src/sink/iceberg/mod.rs [hunk 1]
```diff
@@ -1375,15 +1375,21 @@ pub fn try_matches_arrow_schema(
...
             (ArrowDataType::LargeBinary, ArrowDataType::Binary) => true,
-            (left, right) => left == right,
+            // cases where left != right (metadata, field name mismatch)
+            //
+            // all nested types: in iceberg `field_id` will always be present, but RW doesn't have it:
+            // {"PARQUET:field_id": ".."}
+            //
+            // map: The standard name in arrow is "entries", "key", "value".
+            // in iceberg-rs, it's called "key_value"
+            (left, right) => left.equals_datatype(right),
```
Comment (chenzl25) [inline src/connector/src/sink/iceberg/mod.rs:1385]: Comparing only the datatype. That's exactly what we need!

File: e2e_test/iceberg/test_case/no_partition_append_only.toml [hunk 2]
```diff
@@ -13,24 +13,27 @@ init_sqls = [
     v_date date,
     v_timestamp timestamp,
     v_ts_ntz timestamp_ntz,
-    v_decimal decimal(10,5)
+    v_decimal decimal(10,5),
+    v_map map<int, int>,
+    v_array array<int>,
+    v_struct struct<a:int, b:int>
```
Comment (chenzl25) [inline e2e_test/iceberg/test_case/no_partition_append_only.toml:19]: Have you ever tried nesting a nested type inside another nested type?

**Review by chenzl25 — APPROVED** [review]
Great!

File: e2e_test/iceberg/test_case/no_partition_append_only.toml (hunk 2 above)
Comment (xxchan) [reply e2e_test/iceberg/test_case/no_partition_append_only.toml:19]: Just tried, and found a new icelake bug. It will fail to parse table metadata 🥵

Comment (xxchan) [reply e2e_test/iceberg/test_case/no_partition_append_only.toml:19]: Let's fix this later

**Comment by xxchan** [comment]
ping @risingwavelabs/cargo-lock

**Review by BugenZhao — APPROVED** [review]

**Pull request merged**